File: scrabble-cheater/scrabble_cheater.rb
# Download from:
# https://raw.githubusercontent.com/jonbcard/scrabble-bot/master/src/dictionary.txt
DICT = 'dictionary.txt'
#DICT= '/usr/share/dict/words'

input_chars = if ARGV.empty?
  7.times.map { ('a'..'z').to_a.sample }
else
  ARGV.first.each_char.to_a
end

words = File.read(DICT).downcase.split("\n")

permutations = (1..input_chars.size).to_a.
  flat_map { |n| input_chars.permutation(n).to_a.map(&:join) }

matching_words = words & permutations

puts "Using input characters: #{input_chars.join(", ")}"
puts

if matching_words.any?
  puts "Found matching words:"
  puts matching_words.join("\n")
else
  puts "No matching words. :("
end

File: scrabble-cheater/scrabble_cheater.hs
import System.Environment
import Data.Char
import Data.List

getInputChars :: [String] -> String
getInputChars [] = "tacocat"
getInputChars (x:xs) = x

matchingWords :: String -> [String] -> [String]
matchingWords chars words = filter (\w -> null (w \\ chars)) words

matchingWordsFromDict :: String -> FilePath -> IO [String]
matchingWordsFromDict chars dict = do
  words <- readFile dict
  return (matchingWords chars (lines (map toLower words)))

main = do
  input <- getArgs
  let input_chars = getInputChars input

  putStr "Using input characters: "
  putStrLn $ intersperse ',' input_chars
  putStrLn ""

  --raw_words <- readFile "dictionary.txt"
  --let words = lines $ map toLower raw_words

  matching_words <- matchingWordsFromDict input_chars "dictionary.txt"

  putStrLn "Found matching words:"
  --mapM_ putStrLn (matchingWords input_chars words)
  mapM_ putStrLn matching_words
